File: identity-service/src/main/java/com/stancloud/identityservice/service/AuthService.java
package com.stancloud.identityservice.service;

import com.stancloud.identityservice.config.JwtService;
import com.stancloud.identityservice.dto.UserDto;
import com.stancloud.identityservice.dto.UserResponseDto;
import com.stancloud.identityservice.entity.UserCredential;
import com.stancloud.identityservice.repo.UserCredentialRepo;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class AuthService {
  @Autowired
  private UserCredentialRepo userRepo;
  @Autowired
  private PasswordEncoder passwordEncoder;
  @Autowired
  private AuthenticationManager authenticationManager;

  @Autowired
  private JwtService jwtService;

  public UserResponseDto saveUser(UserDto userDto) {
    UserCredential user = UserCredential.builder()
      .email(userDto.getEmail())
      .password(passwordEncoder.encode(userDto.getPassword()))
      .userRoles(userDto.getUserRoles())
      .build();

    UserCredential savedUser = userRepo.save(user);

    return UserResponseDto.builder()
      .id(savedUser.getId())
      .email(savedUser.getEmail())
      .userRoles(savedUser.getUserRoles())
      .build();
  }

  public void validateToken(String token) {
    jwtService.validateToken(token);
  }

  public String login(UserDto userDto) {
    UserCredential user = userRepo.findByEmail(userDto.getEmail())
      .orElseThrow(() -> new RuntimeException("User not found"));

    if (passwordEncoder.matches(userDto.getPassword(), user.getPassword())) {
      String token = jwtService.generateToken(user);
      return token;
    } else {
      throw new RuntimeException("Username Password incorrect");
    }
//    authenticationManager.authenticate(
//        new UsernamePasswordAuthenticationToken(userDto.getEmail(), userDto.getPassword())
//    );
//
//    UserCredential user = userRepo.findByEmail(userDto.getEmail())
//        .orElseThrow(() -> new RuntimeException("User not found"));
//
//    return jwtService.generateToken(user.getEmail());
  }
}

File: authentication-service/src/main/kotlin/com/stancloud/authenticationservice/auth/AuthService.kt
package com.stancloud.authenticationservice.auth

import com.stancloud.authenticationservice.dto.LoginRequest
import com.stancloud.authenticationservice.dto.LoginResponse
import com.stancloud.authenticationservice.dto.SignupRequest
import com.stancloud.authenticationservice.dto.SignupResponse
import com.stancloud.authenticationservice.config.JwtService
import com.stancloud.authenticationservice.model.User
import com.stancloud.authenticationservice.user.UserRepository
import com.stancloud.authenticationservice.user.UserRole
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service

@Service
class AuthService(
  private val userRepository: UserRepository,
  private val passwordEncoder: PasswordEncoder,
  private val authenticationManager: AuthenticationManager,
  private val jwtTokenProvider: JwtService,
) {

  fun signup(
    request: SignupRequest,
  ): SignupResponse {
  val user: User = User(
    request.firstName,
    request.lastName,
    request.email,
    passwordEncoder.encode(request.password),
    request.roles ?: setOf(UserRole.ROLE_USER),
  )

    return SignupResponse(this.userRepository.save(user))
  }

  fun login(
    request: LoginRequest,
  ): LoginResponse {
    authenticationManager.authenticate(
      UsernamePasswordAuthenticationToken(request.email, request.password)
    )

    val user: User = userRepository.findByEmail(request.email)?: throw Exception("Username Password incorrect")

    return LoginResponse(jwtTokenProvider.generateToken(user))
  }

  fun verify(token: String): Boolean {
    return jwtTokenProvider.validateToken(token)
  }
}
